### PR TITLE
adapter: Update privileges when an owner is updated

### PR DESCRIPTION
**Pull request by jkosh44**
This commit will update the privilege of an object whenever the
object's owner is updated, in the following ways:

  - All privileges granted by the old owner are updated so that their
  grantor is the new owner.
  - All privileges granted to the old owner are transferred to the new
  owner.

Part of #11579

### Motivation
This PR adds a known-desirable feature.

### Tips for Reviewer
- Hide whitespace and the diff becomes much smaller.
- This builds off of https://github.com/MaterializeInc/materialize/pull/18700, only the final commit is unique to this PR. Ideally I'd like to get them in the same release otherwise we may need to write a migration. Almost everything in here is pulled directly from PostgreSQL so hopefully it isn't too controversial.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by jkosh44** [comment]
This builds off of https://github.com/MaterializeInc/materialize/pull/18700. Ideally I'd like to get them in the same release otherwise we may need to write a migration. Almost everything in here is pulled directly from PostgreSQL so hopefully it isn't too controversial.

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -5758,107 +5758,186 @@ impl Catalog {
...
+                        new_owner: RoleId,
+                    ) {
+                        let mut new_present = false;
+                        for privilege in privileges.iter_mut() {
+                            // Old owner's granted privilege are updated to be granted by the new
+                            // owner.
+                            if privilege.grantor == old_owner {
+                                privilege.grantor = new_owner;
+                            } else if privilege.grantor == new_owner {
+                                new_present = true;
```
Comment (def-) [inline src/adapter/src/catalog.rs:5777]: This seems untested

File: src/adapter/src/catalog.rs [hunk 2]
```diff
@@ -5758,107 +5758,186 @@ impl Catalog {
...
+                            }
+                            // Old owner's privileges is given to the new owner.
+                            if privilege.grantee == old_owner {
+                                privilege.grantee = new_owner;
+                            } else if privilege.grantee == new_owner {
+                                new_present = true;
```
Comment (def-) [inline src/adapter/src/catalog.rs:5783]: This seems untested

File: src/adapter/src/catalog.rs [hunk 3]
```diff
@@ -5758,107 +5758,186 @@ impl Catalog {
...
+                            }
+                        }
+
+                        // If the old privilege list contained references to the new owner, we may
+                        // have created duplicate entries. Here we try and consolidate them. This
+                        // is inspired by PostgreSQL's algorithm but not identical.
+                        if new_present {
+                            // Group privileges by (grantee, grantor).
+                            let privilege_map: BTreeMap<_, Vec<_>> = privileges.into_iter().fold(
```
Comment (def-) [inline src/adapter/src/catalog.rs:5792]: This pretty large block seems untested.

File: src/repr/src/adt/mz_acl_item.rs [hunk 4]
```diff
@@ -147,6 +147,14 @@ pub struct MzAclItem {
 }
 
 impl MzAclItem {
+    pub fn empty(grantee: RoleId, grantor: RoleId) -> MzAclItem {
```
Comment (def-) [inline src/repr/src/adt/mz_acl_item.rs:150]: Is this function tested anywhere?

**Review by def- — COMMENTED** [review]
Sorry for the late review.

I tried out code coverage on this PR: https://buildkite.com/materialize/coverage/builds/26#0187900b-9c08-441b-be29-a36e2d86a3c1
It found some of the added lines were not run in any test, would it make sense and be possible to test them?

**Pull request merged**